pkg/cloudprovider/providers/vsphere/vsphere_util.go

```diff
@@ -316,45 +319,54 @@ func getDatastoresForZone(ctx context.Context, dc *vclib.Datacenter, nodeManager
-	pbmClient, err := vclib.NewPbmClient(ctx, dc.Client())
+func getPbmCompatibleZonedDatastore(ctx context.Context, vcClient *vim25.Client, storagePolicyName string, zonedDatastores []*vclib.DatastoreInfo) (*vclib.DatastoreInfo, error) {
+	pbmClient, err := vclib.NewPbmClient(ctx, vcClient)
 	if err != nil {
-		return "", err
+		return nil, err
 	}
 	storagePolicyID, err := pbmClient.ProfileIDByName(ctx, storagePolicyName)
 	if err != nil {
 		klog.Errorf("Failed to get Profile ID by name: %s. err: %+v", storagePolicyName, err)
-		return "", err
+		return nil, err
 	}
-	compatibleDatastores, _, err := pbmClient.GetCompatibleDatastores(ctx, dc, storagePolicyID, zonedDatastores)
+	compatibleDatastores, _, err := pbmClient.GetCompatibleDatastores(ctx, storagePolicyID, zonedDatastores)
 	if err != nil {
 		klog.Errorf("Failed to get compatible datastores from datastores : %+v with storagePolicy: %s. err: %+v",
 			zonedDatastores, storagePolicyID, err)
-		return "", err
+		return nil, err
 	}
 	klog.V(9).Infof("compatibleDatastores : %+v", compatibleDatastores)
-	datastore, err := getMostFreeDatastoreName(ctx, dc.Client(), compatibleDatastores)
+	datastore, err := getMostFreeDatastore(ctx, vcClient, compatibleDatastores)
 	if err != nil {
 		klog.Errorf("Failed to get most free datastore from compatible datastores: %+v. err: %+v", compatibleDatastores, err)
-		return "", err
+		return nil, err
 	}
-	klog.V(4).Infof("Most free datastore : %+s", datastore)
+	klog.V(4).Infof("Most free datastore : %+s", datastore.Info.Name)
 	return datastore, err
 }
 
-func (vs *VSphere) setVMOptions(ctx context.Context, dc *vclib.Datacenter, resourcePoolPath string) (*vclib.VMOptions, error) {
+func (vs *VSphere) setVMOptions(ctx context.Context, connection *vclib.VSphereConnection, ds *vclib.Datastore) (*vclib.VMOptions, error) {
 	var vmOptions vclib.VMOptions
-	resourcePool, err := dc.GetResourcePool(ctx, resourcePoolPath)
+	dsHosts, err := ds.GetDatastoreHostMounts(ctx)
+	if err != nil {
+		klog.Errorf("Failed to get datastore host mounts for %v: %+v", ds, err)
+		return nil, err
+	}
+	// pick a host at random to use for Volume creation
+	dsHostMoref := dsHosts[rand.New(rand.NewSource(time.Now().UnixNano())).Intn(len(dsHosts))]
+	dummyVMHost := object.NewHostSystem(connection.Client, dsHostMoref)
+	resourcePool, err := dummyVMHost.ResourcePool(ctx)
 	if err != nil {
+		klog.Errorf("Failed to get resource pool from host %v", dummyVMHost)
 		return nil, err
 	}
-	klog.V(9).Infof("Resource pool path %s, resourcePool %+v", resourcePoolPath, resourcePool)
-	folder, err := dc.GetFolderByPath(ctx, vs.cfg.Workspace.Folder)
+	folder, err := ds.Datacenter.GetFolderByPath(ctx, vs.cfg.Workspace.Folder)
 	if err != nil {
 		return nil, err
 	}
 	vmOptions.VMFolder = folder
 	vmOptions.VMResourcePool = resourcePool
+	vmOptions.VMHost = dummyVMHost
```

Comment:
Since we already know the folder and host resource pool, do we need dummyVMHost? If no, could you please remove it?

Reply:
Sure will do.